cmd/query/app/apiv3/otlp_translator.go

```diff
@@ -0,0 +1,364 @@
+	"fmt"
+	"hash/fnv"
+	"math"
+	"sort"
+	"strconv"
+	"strings"
+
+	"go.opentelemetry.io/collector/translator/conventions"
+	tracetranslator "go.opentelemetry.io/collector/translator/trace"
+
+	"github.com/jaegertracing/jaeger/model"
+	commonv1 "github.com/jaegertracing/jaeger/proto-gen/otel/common/v1"
+	resourcev1 "github.com/jaegertracing/jaeger/proto-gen/otel/resource/v1"
+	v1 "github.com/jaegertracing/jaeger/proto-gen/otel/trace/v1"
+)
+
+// OpenTelemetry collector implements translator from Jaeger model to pdata (wrapper around OTLP).
+// However, it cannot be used because the imported OTLP in the translator is in the collector's private package.
+func jaegerSpansToOTLP(spans []*model.Span) []*v1.ResourceSpans {
+	spansByLibrary := make(map[resource]map[instrumentationLibrary]*v1.InstrumentationLibrarySpans)
+	for _, s := range spans {
+		otlpSpan, res, library := jSpanToOTLP(s)
+		resourceSpans, ok := spansByLibrary[res]
+		if !ok {
+			resourceSpans = map[instrumentationLibrary]*v1.InstrumentationLibrarySpans{}
+			resourceSpans[library] = &v1.InstrumentationLibrarySpans{
+				InstrumentationLibrary: &commonv1.InstrumentationLibrary{
+					Name:    library.name,
+					Version: library.version,
+				},
+			}
+			spansByLibrary[res] = resourceSpans
+		}
+		resourceSpans[library].Spans = append(resourceSpans[library].GetSpans(), otlpSpan)
+	}
+
+	var rss []*v1.ResourceSpans
+	for res, libMap := range spansByLibrary {
+		rs := &v1.ResourceSpans{
+			Resource: res.resource,
+		}
+		for _, v := range libMap {
+			rs.InstrumentationLibrarySpans = append(rs.InstrumentationLibrarySpans, v)
+		}
+		rss = append(rss, rs)
+	}
+	return rss
+}
+
+type instrumentationLibrary struct {
+	name, version string
+}
+
+// helper type used as a map key
+type resource struct {
+	serviceName string
+	// concatenated and hashed string tags
+	// to make sure services are uniquely grouped
+	tagsHash uint32
+	resource *resourcev1.Resource
+}
+
+func jSpanToOTLP(jSpan *model.Span) (*v1.Span, resource, instrumentationLibrary) {
+	tags := model.KeyValues(jSpan.GetTags())
+	status, ignoreKeys := getSpanStatus(tags)
+	if ignoreKeys == nil {
+		ignoreKeys = map[string]bool{}
+	}
+
+	traceState := getTraceStateFromAttrs(tags)
+	if traceState != "" {
+		ignoreKeys[tracetranslator.TagW3CTraceState] = true
+	}
+
+	s := &v1.Span{
+		TraceId:           uint64ToTraceID(jSpan.TraceID.High, jSpan.TraceID.Low),
+		SpanId:            uint64ToSpanID(uint64(jSpan.SpanID)),
+		ParentSpanId:      uint64ToSpanID(uint64(jSpan.ParentSpanID())),
+		TraceState:        traceState,
+		Name:              jSpan.GetOperationName(),
+		StartTimeUnixNano: uint64(jSpan.GetStartTime().UnixNano()),
+		EndTimeUnixNano:   uint64(jSpan.GetStartTime().Add(jSpan.GetDuration()).UnixNano()),
+		Events:            jLogsToOTLP(jSpan.GetLogs()),
+		Links:             jReferencesToOTLP(jSpan.GetReferences(), jSpan.ParentSpanID()),
+		Status:            status,
+	}
+	if kind, found := jSpan.GetSpanKind(); found {
```

Comment:
If span kind isn't found, `s.Kind` defaults to the zero value of `0` == `SPAN_KIND_UNSPECIFIED`. But comment reads:
```
	// Unspecified. Do NOT use as default.
	// Implementations MAY assume SpanKind to be INTERNAL when receiving UNSPECIFIED.
	Span_SPAN_KIND_UNSPECIFIED Span_SpanKind = 0
```

Should we be setting a default of `SPAN_KIND_INTERNAL` when constructing the span first?